File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/SettingsFragment.java
package com.github.dedis.popstellar.ui.home;

import android.os.Bundle;
import android.view.*;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.SwitchPreference;

import com.github.dedis.popstellar.R;
import com.github.dedis.popstellar.utility.MessageValidator;
import com.github.dedis.popstellar.utility.NetworkLogger;
import com.takisoft.preferencex.EditTextPreference;
import com.takisoft.preferencex.PreferenceFragmentCompat;

import java.util.Objects;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class SettingsFragment extends PreferenceFragmentCompat {

  public static final String TAG = SettingsFragment.class.getSimpleName();

  public static SettingsFragment newInstance() {
    return new SettingsFragment();
  }

  private SettingsViewModel settingsViewModel;
  private HomeViewModel homeViewModel;

  public SettingsFragment() {
    // Empty constructor
  }

  @Override
  public void onCreatePreferencesFix(@Nullable Bundle savedInstanceState, String rootKey) {
    // Load the preferences from an XML resource
    setPreferencesFromResource(R.xml.settings_preferences, rootKey);

    // Set callbacks for debugging section
    setDebuggingPreferences();
  }

  @Override
  public View onCreateView(
      @NonNull LayoutInflater inflater,
      @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {

    homeViewModel = HomeActivity.obtainViewModel(requireActivity());
    settingsViewModel = HomeActivity.obtainSettingsViewModel(requireActivity());

    return super.onCreateView(inflater, container, savedInstanceState);
  }

  @Override
  public void onResume() {
    super.onResume();
    homeViewModel.setPageTitle(R.string.settings);
    homeViewModel.setIsHome(false);
  }

  /** Function that sets the callback for switching the preferences in the section "Debugging" */
  private void setDebuggingPreferences() {
    EditTextPreference serverUrl =
        getPreferenceManager().findPreference(getString(R.string.settings_server_url_key));
    SwitchPreference enableLogging =
        getPreferenceManager().findPreference(getString(R.string.settings_logging_key));

    // Initially the switch is disabled unless it's already on
    Objects.requireNonNull(enableLogging).setEnabled(enableLogging.isChecked());

    // Initially the edit text is enabled unless the logging is already on
    Objects.requireNonNull(serverUrl).setEnabled(!enableLogging.isChecked());

    // Set the callback for managing the logging
    enableLogging.setOnPreferenceChangeListener(
        ((preference, newValue) -> {
          // Save the URL preference if the switch is turned on
          boolean isLoggingEnabled = (Boolean) newValue;

          if (isLoggingEnabled) {
            settingsViewModel.enableLogging();
          } else {
            settingsViewModel.disableLogging();
          }
          // Enable the edit text if switch off, disable it if on
          // (i.e. the switch has to be off to change the server address)
          serverUrl.setEnabled(!isLoggingEnabled);

          // Return true to allow the preference to be saved
          return true;
        }));

    // Set the callback for managing the server url
    serverUrl.setOnPreferenceChangeListener(
        (preference, newValue) -> {
          // Check if the new value is a valid URL
          String newUrl = (String) newValue;

          try {
            MessageValidator.verify().validUrl(newUrl);
            // Save the server url
            NetworkLogger.setServerUrl(newUrl);
            // Enable the switch preference based on URL validity
            enableLogging.setEnabled(true);
            return true;
          } catch (IllegalArgumentException e) {
            enableLogging.setEnabled(false);
            // Show an error message and prevent the preference from being updated
            Toast.makeText(getContext(), R.string.error_settings_url_server, Toast.LENGTH_SHORT)
                .show();
            return true;
          }
        });
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/SettingsFragment.kt
package com.github.dedis.popstellar.ui.home

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.preference.Preference
import androidx.preference.SwitchPreference
import com.github.dedis.popstellar.R
import com.github.dedis.popstellar.utility.MessageValidator.verify
import com.github.dedis.popstellar.utility.NetworkLogger.Companion.setServerUrl
import com.takisoft.preferencex.EditTextPreference
import com.takisoft.preferencex.PreferenceFragmentCompat
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class SettingsFragment : PreferenceFragmentCompat() {
  private lateinit var settingsViewModel: SettingsViewModel
  private lateinit var homeViewModel: HomeViewModel

  override fun onCreatePreferencesFix(savedInstanceState: Bundle?, rootKey: String?) {
    // Load the preferences from an XML resource
    setPreferencesFromResource(R.xml.settings_preferences, rootKey)

    // Set callbacks for debugging section
    setDebuggingPreferences()
  }

  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
  ): View {
    homeViewModel = HomeActivity.obtainViewModel(requireActivity())
    settingsViewModel = HomeActivity.obtainSettingsViewModel(requireActivity())

    handleBackNav()

    return super.onCreateView(inflater, container, savedInstanceState)
  }

  override fun onResume() {
    super.onResume()
    homeViewModel.setPageTitle(R.string.settings)
    homeViewModel.setIsHome(false)
  }

  /** Function that sets the callback for switching the preferences in the section "Debugging" */
  private fun setDebuggingPreferences() {
    val serverUrl =
        preferenceManager.findPreference<EditTextPreference>(
            getString(R.string.settings_server_url_key))
    val enableLogging =
        preferenceManager.findPreference<SwitchPreference>(getString(R.string.settings_logging_key))

    // Initially the switch is disabled unless it's already on
    enableLogging!!.isEnabled = enableLogging.isChecked

    // Initially the edit text is enabled unless the logging is already on
    serverUrl!!.isEnabled = !enableLogging.isChecked

    // Set the callback for managing the logging
    enableLogging.onPreferenceChangeListener =
        Preference.OnPreferenceChangeListener { _: Preference?, newValue: Any ->
          // Save the URL preference if the switch is turned on
          val isLoggingEnabled = newValue as Boolean
          if (isLoggingEnabled) {
            settingsViewModel.enableLogging()
          } else {
            settingsViewModel.disableLogging()
          }
          // Enable the edit text if switch off, disable it if on
          // (i.e. the switch has to be off to change the server address)
          serverUrl.isEnabled = !isLoggingEnabled
          true
        }

    // Set the callback for managing the server url
    serverUrl.onPreferenceChangeListener =
        Preference.OnPreferenceChangeListener returnPoint@{ _: Preference?, newValue: Any ->
          val newUrl = newValue as String
          try {
            // Check if the new value is a valid URL
            verify().validUrl(newUrl)

            // Save the server url
            setServerUrl(newUrl)

            // Enable the switch preference based on URL validity
            enableLogging.isEnabled = true

            return@returnPoint true
          } catch (e: IllegalArgumentException) {
            enableLogging.isEnabled = false
            // Show an error message and prevent the preference from being updated
            Toast.makeText(context, R.string.error_settings_url_server, Toast.LENGTH_SHORT).show()
            return@returnPoint true
          }
        }
  }

  private fun handleBackNav() {
    HomeActivity.addBackNavigationCallbackToHome(requireActivity(), viewLifecycleOwner, TAG)
  }

  companion object {
    val TAG: String = SettingsFragment::class.java.simpleName

    @JvmStatic
    fun newInstance(): SettingsFragment {
      return SettingsFragment()
    }
  }
}
